File: lib/api/ic-score.ts
/**
 * IC Score API Client
 *
 * Handles all API requests related to InvestorCenter's proprietary IC Score system.
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8080/api/v1';

/**
 * Factor calculation metadata - shows the inputs used to calculate each factor score
 */
export interface FactorMetadata {
  // Value factor inputs
  pe_ratio?: number;
  pb_ratio?: number;
  ps_ratio?: number;
  pe_score?: number;
  pb_score?: number;
  ps_score?: number;

  // Growth factor inputs
  revenue_growth?: number;
  eps_growth?: number;
  revenue_score?: number;
  eps_score?: number;

  // Profitability factor inputs
  net_margin?: number;
  roe?: number;
  roa?: number;
  margin_score?: number;
  roe_score?: number;
  roa_score?: number;

  // Financial health factor inputs
  debt_to_equity?: number;
  current_ratio?: number;
  de_score?: number;
  cr_score?: number;

  // Momentum factor inputs
  return_1m?: number;
  return_3m?: number;
  return_6m?: number;
  return_12m?: number;

  // Technical factor inputs
  rsi?: number;
  macd_histogram?: number;
  price_vs_sma50?: number;
  price_vs_sma200?: number;

  // Analyst consensus inputs
  buy_count?: number;
  hold_count?: number;
  sell_count?: number;
  consensus_score?: number;
  price_target?: number;
  current_price?: number;
  upside_pct?: number;

  // Insider activity inputs
  net_shares?: number;
  buy_transactions?: number;
  sell_transactions?: number;

  // Institutional inputs
  institution_count?: number;
  total_shares?: number;
  holdings_change_pct?: number;

  // News sentiment inputs
  sentiment_avg?: number;
  article_count?: number;

  // Generic field for any additional data
  [key: string]: number | string | undefined;
}

/**
 * Calculation metadata stored with each IC Score
 */
export interface CalculationMetadata {
  factors: Record<string, FactorMetadata>;
  weights_used: Record<string, number>;
  calculated_at: string;
}

/**
 * IC Score Data Interface
 * Represents the complete IC Score data for a stock
 */
export interface ICScoreData {
  ticker: string;
  date: string;
  overall_score: number;
  value_score: number | null;
  growth_score: number | null;
  profitability_score: number | null;
  financial_health_score: number | null;
  momentum_score: number | null;
  analyst_consensus_score: number | null;
  insider_activity_score: number | null;
  institutional_score: number | null;
  news_sentiment_score: number | null;
  technical_score: number | null;
  rating: string;
  sector_percentile: number | null;
  confidence_level: string;
  data_completeness: number;
  calculated_at: string;
  factor_count: number;
  available_factors: string[];
  missing_factors: string[];
  calculation_metadata?: CalculationMetadata;
}

/**
 * IC Score List Item
 * Simplified view for admin/list displays
 */
export interface ICScoreListItem {
  ticker: string;
  overall_score: number;
  rating: string;
  data_completeness: number;
  calculated_at: string;
}

/**
 * IC Score Factor Details
 * Detailed information about a specific factor
 */
export interface ICScoreFactor {
  name: string;
  display_name: string;
  score: number | null;
  weight: number;
  grade: string;
  available: boolean;
  description: string;
}

/**
 * API Response wrapper
 */
interface APIResponse<T> {
  data: T;
  meta?: Record<string, any>;
}

/**
 * Fetch IC Score for a specific ticker
 *
 * @param ticker Stock symbol (e.g., "AAPL")
 * @returns IC Score data or null if not available
 */
export async function getICScore(ticker: string): Promise<ICScoreData | null> {
  try {
    const response = await fetch(`${API_BASE_URL}/stocks/${ticker.toUpperCase()}/ic-score`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
      cache: 'no-store', // Always fetch fresh data
    });

    if (response.status === 404) {
      // Score not available for this ticker
      return null;
    }

    if (!response.ok) {
      console.error(`Failed to fetch IC Score for ${ticker}: ${response.status}`);
      return null;
    }

    const result: APIResponse<ICScoreData> = await response.json();
    return result.data;
  } catch (error) {
    console.error(`Error fetching IC Score for ${ticker}:`, error);
    return null;
  }
}

/**
 * Fetch IC Score history for a ticker
 *
 * @param ticker Stock symbol
 * @param days Number of days of history to fetch (default: 90)
 * @returns Array of historical IC Scores
 */
export async function getICScoreHistory(
  ticker: string,
  days: number = 90
): Promise<ICScoreData[]> {
  try {
    const response = await fetch(
      `${API_BASE_URL}/stocks/${ticker.toUpperCase()}/ic-score/history?days=${days}`,
      {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
        cache: 'no-store',
      }
    );

    if (!response.ok) {
      console.error(`Failed to fetch IC Score history for ${ticker}: ${response.status}`);
      return [];
    }

    const result: APIResponse<ICScoreData[]> = await response.json();
    return result.data;
  } catch (error) {
    console.error(`Error fetching IC Score history for ${ticker}:`, error);
    return [];
  }
}

/**
 * Fetch all IC Scores with pagination
 *
 * @param params Query parameters for filtering and pagination
 * @returns Paginated list of IC Scores
 */
export async function getICScores(params?: {
  limit?: number;
  offset?: number;
  search?: string;
  sort?: string;
  order?: 'asc' | 'desc';
}): Promise<{
  data: ICScoreListItem[];
  meta: {
    total: number;
    limit: number;
    offset: number;
    total_stocks: number;
    coverage_percent: number;
  };
}> {
  try {
    const queryParams = new URLSearchParams();
    if (params?.limit) queryParams.set('limit', params.limit.toString());
    if (params?.offset) queryParams.set('offset', params.offset.toString());
    if (params?.search) queryParams.set('search', params.search);
    if (params?.sort) queryParams.set('sort', params.sort);
    if (params?.order) queryParams.set('order', params.order);

    const url = `${API_BASE_URL}/ic-scores${queryParams.toString() ? '?' + queryParams.toString() : ''}`;
    const response = await fetch(url, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
      cache: 'no-store',
    });

    if (!response.ok) {
      console.error(`Failed to fetch IC Scores: ${response.status}`);
      return {
        data: [],
        meta: {
          total: 0,
          limit: params?.limit || 20,
          offset: params?.offset || 0,
          total_stocks: 0,
          coverage_percent: 0,
        },
      };
    }

    return await response.json();
  } catch (error) {
    console.error('Error fetching IC Scores:', error);
    return {
      data: [],
      meta: {
        total: 0,
        limit: params?.limit || 20,
        offset: params?.offset || 0,
        total_stocks: 0,
        coverage_percent: 0,
      },
    };
  }
}

/**
 * Get letter grade for a score
 *
 * Aligned with IC Score rating system:
 * - A range (80-100): Strong Buy
 * - B range (65-79): Buy
 * - C range (50-64): Hold
 * - D range (35-49): Underperform
 * - F range (0-34): Sell
 *
 * @param score Numeric score (0-100)
 * @returns Letter grade (A+ to F)
 */
export function getLetterGrade(score: number | null): string {
  if (score === null) return 'N/A';

  // A range: Strong Buy (80-100)
  if (score >= 93) return 'A+';
  if (score >= 86) return 'A';
  if (score >= 80) return 'A-';

  // B range: Buy (65-79)
  if (score >= 75) return 'B+';
  if (score >= 70) return 'B';
  if (score >= 65) return 'B-';

  // C range: Hold (50-64)
  if (score >= 60) return 'C+';
  if (score >= 55) return 'C';
  if (score >= 50) return 'C-';

  // D range: Underperform (35-49)
  if (score >= 45) return 'D+';
  if (score >= 40) return 'D';
  if (score >= 35) return 'D-';

  // F range: Sell (0-34)
  return 'F';
}

/**
 * Get rating label from score
 */
export function getRating(score: number | null): string {
  if (score === null) return 'N/A';
  if (score >= 80) return 'Strong Buy';
  if (score >= 65) return 'Buy';
  if (score >= 50) return 'Hold';
  if (score >= 35) return 'Underperform';
  return 'Sell';
}

/**
 * Get grade tier description
 */
export function getGradeTier(score: number | null): { letter: string; rating: string; description: string } {
  if (score === null) return { letter: 'N/A', rating: 'N/A', description: 'Data not available' };

  if (score >= 80) return {
    letter: getLetterGrade(score),
    rating: 'Strong Buy',
    description: 'Exceptional investment profile'
  };
  if (score >= 65) return {
    letter: getLetterGrade(score),
    rating: 'Buy',
    description: 'Favorable investment profile'
  };
  if (score >= 50) return {
    letter: getLetterGrade(score),
    rating: 'Hold',
    description: 'Neutral investment profile'
  };
  if (score >= 35) return {
    letter: getLetterGrade(score),
    rating: 'Underperform',
    description: 'Below average investment profile'
  };
  return {
    letter: getLetterGrade(score),
    rating: 'Sell',
    description: 'Weak investment profile'
  };
}

/**
 * Get color class for a score
 *
 * @param score Numeric score (0-100)
 * @returns Tailwind color class
 */
export function getScoreColor(score: number | null): string {
  if (score === null) return 'text-ic-text-muted';

  if (score >= 80) return 'text-green-600';
  if (score >= 65) return 'text-green-500';
  if (score >= 50) return 'text-yellow-500';
  if (score >= 35) return 'text-orange-500';
  return 'text-red-500';
}

/**
 * Get background color class for a score
 */
export function getScoreBgColor(score: number | null): string {
  if (score === null) return 'bg-ic-surface';

  if (score >= 80) return 'bg-green-100';
  if (score >= 65) return 'bg-green-50';
  if (score >= 50) return 'bg-yellow-50';
  if (score >= 35) return 'bg-orange-50';
  return 'bg-red-50';
}

/**
 * Get all factor details with metadata
 */
export function getFactorDetails(icScore: ICScoreData): ICScoreFactor[] {
  const factors: ICScoreFactor[] = [
    {
      name: 'value',
      display_name: 'Value',
      score: icScore.value_score,
      weight: 12,
      grade: getLetterGrade(icScore.value_score),
      available: icScore.value_score !== null,
      description: 'P/E, P/B, P/S ratios vs sector',
    },
    {
      name: 'growth',
      display_name: 'Growth',
      score: icScore.growth_score,
      weight: 15,
      grade: getLetterGrade(icScore.growth_score),
      available: icScore.growth_score !== null,
      description: 'Revenue, EPS growth trends',
    },
    {
      name: 'profitability',
      display_name: 'Profitability',
      score: icScore.profitability_score,
      weight: 12,
      grade: getLetterGrade(icScore.profitability_score),
      available: icScore.profitability_score !== null,
      description: 'Margins, ROE, ROA',
    },
    {
      name: 'financial_health',
      display_name: 'Financial Health',
      score: icScore.financial_health_score,
      weight: 10,
      grade: getLetterGrade(icScore.financial_health_score),
      available: icScore.financial_health_score !== null,
      description: 'Debt ratios, liquidity',
    },
    {
      name: 'momentum',
      display_name: 'Momentum',
      score: icScore.momentum_score,
      weight: 8,
      grade: getLetterGrade(icScore.momentum_score),
      available: icScore.momentum_score !== null,
      description: 'Price trends, relative strength',
    },
    {
      name: 'analyst_consensus',
      display_name: 'Analyst Consensus',
      score: icScore.analyst_consensus_score,
      weight: 10,
      grade: getLetterGrade(icScore.analyst_consensus_score),
      available: icScore.analyst_consensus_score !== null,
      description: 'Buy/sell ratings, price targets',
    },
    {
      name: 'insider_activity',
      display_name: 'Insider Activity',
      score: icScore.insider_activity_score,
      weight: 8,
      grade: getLetterGrade(icScore.insider_activity_score),
      available: icScore.insider_activity_score !== null,
      description: 'Insider buying/selling trends',
    },
    {
      name: 'institutional',
      display_name: 'Institutional',
      score: icScore.institutional_score,
      weight: 10,
      grade: getLetterGrade(icScore.institutional_score),
      available: icScore.institutional_score !== null,
      description: 'Institutional ownership changes',
    },
    {
      name: 'news_sentiment',
      display_name: 'News Sentiment',
      score: icScore.news_sentiment_score,
      weight: 7,
      grade: getLetterGrade(icScore.news_sentiment_score),
      available: icScore.news_sentiment_score !== null,
      description: 'News analysis and sentiment',
    },
    {
      name: 'technical',
      display_name: 'Technical',
      score: icScore.technical_score,
      weight: 8,
      grade: getLetterGrade(icScore.technical_score),
      available: icScore.technical_score !== null,
      description: 'RSI, MACD, trend indicators',
    },
  ];

  return factors;
}

/**
 * Get star rating (0-5 stars) based on score
 */
export function getStarRating(score: number): number {
  return Math.round((score / 100) * 5);
}

File: backend/models/ic_score.go
package models

import (
	"time"

	"github.com/shopspring/decimal"
)

// LifecycleStage represents the company lifecycle classification
type LifecycleStage string

const (
	LifecycleHypergrowth LifecycleStage = "hypergrowth"
	LifecycleGrowth      LifecycleStage = "growth"
	LifecycleMature      LifecycleStage = "mature"
	LifecycleValue       LifecycleStage = "value"
	LifecycleTurnaround  LifecycleStage = "turnaround"
)

// ICScore represents the InvestorCenter proprietary IC Score
// v2.1: Added lifecycle_stage, sector context, and smoothing fields
// Phase 2: Added earnings_revisions, historical_value, dividend_quality
type ICScore struct {
	ID                    int64            `json:"id" db:"id"`
	Ticker                string           `json:"ticker" db:"ticker"`
	Date                  time.Time        `json:"date" db:"date"`
	OverallScore          decimal.Decimal  `json:"overall_score" db:"overall_score"`
	ValueScore            *decimal.Decimal `json:"value_score" db:"value_score"`
	GrowthScore           *decimal.Decimal `json:"growth_score" db:"growth_score"`
	ProfitabilityScore    *decimal.Decimal `json:"profitability_score" db:"profitability_score"`
	FinancialHealthScore  *decimal.Decimal `json:"financial_health_score" db:"financial_health_score"`
	MomentumScore         *decimal.Decimal `json:"momentum_score" db:"momentum_score"`
	AnalystConsensusScore *decimal.Decimal `json:"analyst_consensus_score" db:"analyst_consensus_score"`
	InsiderActivityScore  *decimal.Decimal `json:"insider_activity_score" db:"insider_activity_score"`
	InstitutionalScore    *decimal.Decimal `json:"institutional_score" db:"institutional_score"`
	NewsSentimentScore    *decimal.Decimal `json:"news_sentiment_score" db:"news_sentiment_score"`
	TechnicalScore        *decimal.Decimal `json:"technical_score" db:"technical_score"`

	// Phase 2: New factor scores
	EarningsRevisionsScore *decimal.Decimal `json:"earnings_revisions_score" db:"earnings_revisions_score"`
	HistoricalValueScore   *decimal.Decimal `json:"historical_value_score" db:"historical_value_score"`
	DividendQualityScore   *decimal.Decimal `json:"dividend_quality_score" db:"dividend_quality_score"`

	Rating           *string          `json:"rating" db:"rating"`
	SectorPercentile *decimal.Decimal `json:"sector_percentile" db:"sector_percentile"`
	ConfidenceLevel  *string          `json:"confidence_level" db:"confidence_level"`
	DataCompleteness *decimal.Decimal `json:"data_completeness" db:"data_completeness"`

	// v2.1: Lifecycle and sector context
	LifecycleStage   *string          `json:"lifecycle_stage" db:"lifecycle_stage"`
	RawScore         *decimal.Decimal `json:"raw_score" db:"raw_score"`
	SmoothingApplied bool             `json:"smoothing_applied" db:"smoothing_applied"`
	WeightsUsed      map[string]any   `json:"weights_used,omitempty" db:"weights_used"`
	SectorRank       *int             `json:"sector_rank" db:"sector_rank"`
	SectorTotal      *int             `json:"sector_total" db:"sector_total"`

	CalculationMetadata map[string]any `json:"calculation_metadata,omitempty" db:"calculation_metadata"`
	CalculatedAt        time.Time      `json:"calculated_at" db:"created_at"`
}

// ICScoreResponse represents the API response for IC Score
// v2.1: Added lifecycle_stage, sector_rank, peer_comparison
// Phase 2: Added earnings_revisions, historical_value, dividend_quality
type ICScoreResponse struct {
	Ticker                string   `json:"ticker"`
	Date                  string   `json:"date"`
	OverallScore          float64  `json:"overall_score"`
	ValueScore            *float64 `json:"value_score"`
	GrowthScore           *float64 `json:"growth_score"`
	ProfitabilityScore    *float64 `json:"profitability_score"`
	FinancialHealthScore  *float64 `json:"financial_health_score"`
	MomentumScore         *float64 `json:"momentum_score"`
	AnalystConsensusScore *float64 `json:"analyst_consensus_score"`
	InsiderActivityScore  *float64 `json:"insider_activity_score"`
	InstitutionalScore    *float64 `json:"institutional_score"`
	NewsSentimentScore    *float64 `json:"news_sentiment_score"`
	TechnicalScore        *float64 `json:"technical_score"`

	// Phase 2: New factor scores
	EarningsRevisionsScore *float64 `json:"earnings_revisions_score,omitempty"`
	HistoricalValueScore   *float64 `json:"historical_value_score,omitempty"`
	DividendQualityScore   *float64 `json:"dividend_quality_score,omitempty"`

	Rating           string   `json:"rating"`
	SectorPercentile *float64 `json:"sector_percentile"`
	ConfidenceLevel  string   `json:"confidence_level"`
	DataCompleteness float64  `json:"data_completeness"`
	CalculatedAt     string   `json:"calculated_at"`
	FactorCount      int      `json:"factor_count"`
	AvailableFactors []string `json:"available_factors"`
	MissingFactors   []string `json:"missing_factors"`

	// v2.1: New fields
	LifecycleStage *string  `json:"lifecycle_stage,omitempty"`
	SectorRank     *int     `json:"sector_rank,omitempty"`
	SectorTotal    *int     `json:"sector_total,omitempty"`
	ScoringVersion string   `json:"scoring_version"`
	RawScore       *float64 `json:"raw_score,omitempty"`
	IncomeMode     bool     `json:"income_mode"`
}

// ICScoreListItem represents a summary for the admin list view
type ICScoreListItem struct {
	Ticker           string    `json:"ticker" db:"ticker"`
	OverallScore     float64   `json:"overall_score" db:"overall_score"`
	Rating           string    `json:"rating" db:"rating"`
	DataCompleteness float64   `json:"data_completeness" db:"data_completeness"`
	CalculatedAt     time.Time `json:"calculated_at" db:"created_at"`
}

// ToResponse converts ICScore model to API response format
func (ic *ICScore) ToResponse() ICScoreResponse {
	response := ICScoreResponse{
		Ticker:           ic.Ticker,
		Date:             ic.Date.Format("2006-01-02"),
		OverallScore:     toFloat64(ic.OverallScore),
		Rating:           stringOrDefault(ic.Rating, "N/A"),
		ConfidenceLevel:  stringOrDefault(ic.ConfidenceLevel, "Low"),
		DataCompleteness: decimalToFloat64(ic.DataCompleteness, 0),
		CalculatedAt:     ic.CalculatedAt.Format(time.RFC3339),
		AvailableFactors: []string{},
		MissingFactors:   []string{},
	}

	// Track available and missing factors
	factorCount := 0

	if ic.ValueScore != nil {
		v := toFloat64(*ic.ValueScore)
		response.ValueScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "value")
	} else {
		response.MissingFactors = append(response.MissingFactors, "value")
	}

	if ic.GrowthScore != nil {
		v := toFloat64(*ic.GrowthScore)
		response.GrowthScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "growth")
	} else {
		response.MissingFactors = append(response.MissingFactors, "growth")
	}

	if ic.ProfitabilityScore != nil {
		v := toFloat64(*ic.ProfitabilityScore)
		response.ProfitabilityScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "profitability")
	} else {
		response.MissingFactors = append(response.MissingFactors, "profitability")
	}

	if ic.FinancialHealthScore != nil {
		v := toFloat64(*ic.FinancialHealthScore)
		response.FinancialHealthScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "financial_health")
	} else {
		response.MissingFactors = append(response.MissingFactors, "financial_health")
	}

	if ic.MomentumScore != nil {
		v := toFloat64(*ic.MomentumScore)
		response.MomentumScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "momentum")
	} else {
		response.MissingFactors = append(response.MissingFactors, "momentum")
	}

	if ic.AnalystConsensusScore != nil {
		v := toFloat64(*ic.AnalystConsensusScore)
		response.AnalystConsensusScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "analyst_consensus")
	} else {
		response.MissingFactors = append(response.MissingFactors, "analyst_consensus")
	}

	if ic.InsiderActivityScore != nil {
		v := toFloat64(*ic.InsiderActivityScore)
		response.InsiderActivityScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "insider_activity")
	} else {
		response.MissingFactors = append(response.MissingFactors, "insider_activity")
	}

	if ic.InstitutionalScore != nil {
		v := toFloat64(*ic.InstitutionalScore)
		response.InstitutionalScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "institutional")
	} else {
		response.MissingFactors = append(response.MissingFactors, "institutional")
	}

	if ic.NewsSentimentScore != nil {
		v := toFloat64(*ic.NewsSentimentScore)
		response.NewsSentimentScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "news_sentiment")
	} else {
		response.MissingFactors = append(response.MissingFactors, "news_sentiment")
	}

	if ic.TechnicalScore != nil {
		v := toFloat64(*ic.TechnicalScore)
		response.TechnicalScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "technical")
	} else {
		response.MissingFactors = append(response.MissingFactors, "technical")
	}

	// Phase 2: New factor scores
	if ic.EarningsRevisionsScore != nil {
		v := toFloat64(*ic.EarningsRevisionsScore)
		response.EarningsRevisionsScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "earnings_revisions")
	} else {
		response.MissingFactors = append(response.MissingFactors, "earnings_revisions")
	}

	if ic.HistoricalValueScore != nil {
		v := toFloat64(*ic.HistoricalValueScore)
		response.HistoricalValueScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "historical_value")
	} else {
		response.MissingFactors = append(response.MissingFactors, "historical_value")
	}

	if ic.DividendQualityScore != nil {
		v := toFloat64(*ic.DividendQualityScore)
		response.DividendQualityScore = &v
		factorCount++
		response.AvailableFactors = append(response.AvailableFactors, "dividend_quality")
	}
	// Note: dividend_quality is optional, don't add to missing factors

	response.FactorCount = factorCount

	if ic.SectorPercentile != nil {
		v := toFloat64(*ic.SectorPercentile)
		response.SectorPercentile = &v
	}

	// v2.1: Add lifecycle and sector context
	response.LifecycleStage = ic.LifecycleStage
	response.SectorRank = ic.SectorRank
	response.SectorTotal = ic.SectorTotal

	if ic.RawScore != nil {
		v := toFloat64(*ic.RawScore)
		response.RawScore = &v
	}

	// Determine scoring version
	if ic.LifecycleStage != nil {
		response.ScoringVersion = "2.1"
	} else {
		response.ScoringVersion = "2.0"
	}

	return response
}

// Helper functions
func toFloat64(d decimal.Decimal) float64 {
	f, _ := d.Float64()
	return f
}

func decimalToFloat64(d *decimal.Decimal, defaultVal float64) float64 {
	if d == nil {
		return defaultVal
	}
	f, _ := d.Float64()
	return f
}

func stringOrDefault(s *string, defaultVal string) string {
	if s == nil {
		return defaultVal
	}
	return *s
}
